packages/snaps-utils/src/errors.ts

```diff
@@ -20,3 +32,384 @@ export function getErrorMessage(error: unknown) {
+
+  /**
+   * The error message.
+   *
+   * @returns The error message.
+   */
+  get message() {
+    return this.#message;
+  }
+
+  /**
+   * The error stack.
+   *
+   * @returns The error stack.
+   */
+  get stack() {
+    return this.#stack;
+  }
+
+  /**
+   * Convert the error to a JSON object.
+   *
+   * @returns The JSON object.
+   */
+  toJSON(): SerializedSnapErrorWrapper {
+    const cause = isSnapError(this.#error)
+      ? this.#error.serialize()
+      : serializeCause(this.#error);
+
+    return {
+      code: SNAP_ERROR_WRAPPER_CODE,
+      message: SNAP_ERROR_WRAPPER_MESSAGE,
+      data: {
+        cause,
+      },
+    };
+  }
+
+  /**
+   * Serialize the error to a JSON object. This is called by
+   * `@metamask/rpc-errors` when serializing the error.
+   *
+   * @returns The JSON object.
+   */
+  serialize() {
+    return this.toJSON();
+  }
+}
+
+/**
+ * A generic error which can be thrown by a Snap, without it causing the Snap to
+ * crash.
+ */
+export class SnapError extends Error {
+  readonly #code: number;
+
+  readonly #message: string;
+
+  readonly #data: Record<string, Json>;
+
+  readonly #stack?: string;
+
+  /**
+   * Create a new `SnapError`.
+   *
+   * @param error - The error to create the `SnapError` from. If this is a
+   * `string`, it will be used as the error message. If this is an `Error`, its
+   * `message` property will be used as the error message. If this is a
+   * `JsonRpcError`, its `message` property will be used as the error message
+   * and its `code` property will be used as the error code. Otherwise, the
+   * error will be converted to a string and used as the error message.
+   * @param data - Additional data to include in the error. This will be merged
+   * with the error data, if any.
+   */
+  constructor(
+    error: string | Error | JsonRpcError,
+    data: Record<string, Json> = {},
+  ) {
+    const message = getErrorMessage(error);
+    super(message);
+
+    this.#message = message;
+    this.#code = getErrorCode(error);
+    this.#data = { ...getErrorData(error), ...data };
+    this.#stack = super.stack;
+  }
+
+  /**
+   * The error name.
+   *
+   * @returns The error name.
+   */
+  get name() {
+    return 'SnapError';
+  }
+
+  /**
+   * The error code.
+   *
+   * @returns The error code.
+   */
+  get code() {
+    return this.#code;
+  }
+
+  /**
+   * The error message.
+   *
+   * @returns The error message.
+   */
+  get message() {
+    return this.#message;
+  }
+
+  /**
+   * Additional data for the error.
+   *
+   * @returns Additional data for the error.
+   */
+  get data() {
+    return this.#data;
+  }
+
+  /**
+   * The error stack.
+   *
+   * @returns The error stack.
+   */
+  get stack() {
+    return this.#stack;
+  }
+
+  /**
+   * Convert the error to a JSON object.
+   *
+   * @returns The JSON object.
+   */
+  toJSON(): SerializedSnapError {
+    const data = this.stack
+      ? {
+          ...this.data,
+          stack: this.stack,
```

Comment:
We may want to consider serializing the stack outside of the data. That is what rpc-errors does: https://github.com/MetaMask/rpc-errors/blob/main/src/classes.ts#L65

Reply:
I'm not sure, because `stack` is not a valid JSON-RPC error property according to the specification.

Reply:
No, I agree. But for compatibility with the MetaMask JSON-RPC stack this seems less risky.

Reply:
`data` should allow anything, so should be compatible too?

Reply:
Yeah, but for example with `rpc-errors` the `includeStackTrace` option would be broken for these errors.

There may be other such incompatibilities and so there is increased risk in changing the shape IMO